web/pandas/pdeps/0007-copy-on-write.md

```diff
@@ -0,0 +1,575 @@
+and then finally upgrade to the new major release.
+
+## Implementation
+
+An initial implementation has been
+[merged](https://github.com/pandas-dev/pandas/pull/46958) and is available since pandas
+1.5 (and significantly improved in 2.0). It uses weakrefs to keep track of whether the
+data of a Dataframe/Series are viewing the data of another (pandas) object or are being
+viewed by another object. This way, whenever the series/dataframe gets modified, we can
+check if its data first needs to be copied before mutating it.
+
+To test the implementation and experiment with the new behaviour, you can can
+enable it with the following option:
+
+```python
+>>> pd.options.mode.copy_on_write = True
+```
+
+after importing pandas (or setting the `PANDAS_COPY_ON_WRITE=1` environment variable
+before importing pandas).
+
+## Concrete examples
+
+### Chained assignment
+
+Consider a "classic" case of chained indexing, which was the original motivation for the SettingWithCopy warning:
+
+```python
+>>> df[df['B'] > 4]['B'] = 10
+```
+
+That is roughly equivalent to
+
+```python
+>>> df2 = df[df['B'] > 4]  # Copy under NumPy's rules
+>>> df2['B'] = 10  # Update (the copy) df2, df not changed
+>>> del df2  # All references to df2 are lost, goes out of scope
+```
+
+And so `df` is not modified. For this reason, the SettingWithCopyWarning was introduced.
+
+_With this proposal_, any result of an indexing operation behaves as a copy
+(Copy-on-Write), and thus chained assignment will _never_ work. Given that there is then
+no ambiguity, the idea is to drop the warning.
+
+The above example is a case where chained assignment doesn't work with current pandas.
+But there are of course also patterns with chained assignment that currently _do_ work
+and are used. _With this proposal_, any chained assignment will not work, and so those
+cases will stop working (e.g. the case above but switching the order):
+
+```python
+>>> df['B'][df['B'] > 4] = 10
+# or
+>>> df['B'][0:5] = 10
+```
+
+It might be possible to detect this and raise an error (since it can never do what the
+user intended, an error could be appropriate / helpful).
+
+### Filtered dataframe
+
+A typical example where the current SettingWithCopyWarning becomes annoying is when
+filtering a DataFrame (which always already returns a copy):
+
+```python
+>>> df = pd.DataFrame({"A": [1, 2], "B": [3, 4], "C": [5, 6]})
+>>> df_filtered = df[df["A"] > 1]
+>>> df_filtered["new_column"] = 1
+SettingWithCopyWarning:
+A value is trying to be set on a copy of a slice from a DataFrame.
+Try using .loc[row_indexer,col_indexer] = value instead
+```
+
+If you then modify your filtered dataframe (e.g. adding a column), you get the
+unnecessary SettingWIthCopyWarning (with confusing message). The only way to get rid of
+the warning is by doing a defensive copy (`df_filtered = df[df["A"] > 1].copy()`, which
+results in copying the data twice).
```

Comment:
Copying the data twice in current pandas, or with this proposal?

Reply:
Current pandas, CoW will only copy once

Reply:
OK, but I think this should be clarified.

Reply:
Updated